File: modules/vescape-core/ios/location/GpsCourseDeriver.swift
import Foundation

/// A course the rider is actually travelling, derived per precise fix.
///
/// `sourceTimestamp` is the fix the bearing came from, not the fix it was reported on: a retained
/// course keeps the older timestamp so callers can tell a fresh course from a held one.
internal struct GpsCourse: Equatable {
  let bearingDeg: Double
  let sourceTimestamp: Int64
}

/// Turns raw GPS fixes into a *reliable* course.
///
/// A fix's own bearing is noise while standing still — a parked board spins the puck through every
/// heading — so it is only trusted above a walking pace, falls back to the line between the last two
/// fixes when the receiver reports no bearing, and is otherwise held briefly so a momentary stop at
/// a light does not throw the puck back to north.
///
/// Feed precise fixes only, in time order. Stateful, one instance per location source.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/GpsCourseDeriver.kt `GpsCourseDeriver`
/// @platform-diff `reset()` is iOS-only: `LocationTracker` drops the replay's fixes when a
/// replay session ends and clears the deriver with them, so a recorded course cannot bend the first
/// live one. Android's `LocationTracker` outlives every session and keeps both.
/// @parity /modules/vescape-core/src/index.ts `LocationEvent`
internal final class GpsCourseDeriver {
  /// Below roughly a walking pace the reported bearing is receiver noise.
  static let minSpeedMps = 0.8

  /// Two fixes closer than this describe GPS jitter, not travel.
  static let minDistanceM = 2.0

  /// How long a course survives without a fix that can confirm it.
  static let retentionMs: Int64 = 10_000

  private var previousLatitude: Double?
  private var previousLongitude: Double?
  private var previousTimestamp: Int64?
  private var lastCourse: GpsCourse?

  func derive(
    latitude: Double,
    longitude: Double,
    speedMps: Double?,
    bearingDeg: Double?,
    timestamp: Int64
  ) -> GpsCourse? {
    let course = resolve(
      latitude: latitude,
      longitude: longitude,
      speedMps: speedMps,
      bearingDeg: bearingDeg,
      timestamp: timestamp
    )
    previousLatitude = latitude
    previousLongitude = longitude
    previousTimestamp = timestamp
    lastCourse = course
    return course
  }

  func reset() {
    previousLatitude = nil
    previousLongitude = nil
    previousTimestamp = nil
    lastCourse = nil
  }

  private func resolve(
    latitude: Double,
    longitude: Double,
    speedMps: Double?,
    bearingDeg: Double?,
    timestamp: Int64
  ) -> GpsCourse? {
    // A missing speed is treated as moving: a receiver that reports no speed at all should still
    // steer the puck rather than never producing a course.
    let moving = speedMps == nil || (speedMps ?? 0) >= Self.minSpeedMps
    if moving {
      if let reported = GeoMath.normalizeBearingDeg(bearingDeg) {
        return GpsCourse(bearingDeg: reported, sourceTimestamp: timestamp)
      }
      if let derived = derivedBearingDeg(latitude: latitude, longitude: longitude, timestamp: timestamp) {
        return GpsCourse(bearingDeg: derived, sourceTimestamp: timestamp)
      }
    }

    guard let retained = lastCourse else { return nil }
    guard timestamp - retained.sourceTimestamp <= Self.retentionMs else { return nil }
    return retained
  }

  private func derivedBearingDeg(latitude: Double, longitude: Double, timestamp: Int64) -> Double? {
    guard
      let fromLatitude = previousLatitude,
      let fromLongitude = previousLongitude,
      let fromTimestamp = previousTimestamp,
      fromTimestamp < timestamp,
      GeoMath.distanceMeters(fromLatitude, fromLongitude, latitude, longitude) >= Self.minDistanceM
    else { return nil }
    return GeoMath.travelBearingDeg(fromLatitude, fromLongitude, latitude, longitude)
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/GpsCourseDeriver.kt
package expo.modules.vescapecore.location

import expo.modules.vescapecore.geo.GeoMath

/**
 * A course the rider is actually travelling, derived per precise fix.
 *
 * `sourceTimestamp` is the fix the bearing came from, not the fix it was reported on: a retained
 * course keeps the older timestamp so callers can tell a fresh course from a held one.
 */
internal data class GpsCourse(
  val bearingDeg: Double,
  val sourceTimestamp: Long,
)

/**
 * Turns raw GPS fixes into a *reliable* course.
 *
 * A fix's own bearing is noise while standing still — a parked board spins the puck through every
 * heading — so it is only trusted above a walking pace, falls back to the line between the last two
 * fixes when the receiver reports no bearing, and is otherwise held briefly so a momentary stop at
 * a light does not throw the puck back to north.
 *
 * Feed precise fixes only, in time order. Stateful, one instance per location source.
 *
 * @parity /modules/vescape-core/ios/location/GpsCourseDeriver.swift `GpsCourseDeriver`
 * @platform-diff No `reset()`: Android's `LocationTracker` outlives every session and keeps its
 * latest fixes across disconnects, so there is no teardown to hang one off. iOS clears its fix
 * state on disconnect and resets the deriver with it.
 * @parity /modules/vescape-core/src/index.ts `LocationEvent`
 */
internal class GpsCourseDeriver {
  private var previousLatitude: Double? = null
  private var previousLongitude: Double? = null
  private var previousTimestamp: Long? = null
  private var lastCourse: GpsCourse? = null

  fun derive(
    latitude: Double,
    longitude: Double,
    speedMps: Double?,
    bearingDeg: Double?,
    timestamp: Long,
  ): GpsCourse? {
    val course = resolve(latitude, longitude, speedMps, bearingDeg, timestamp)
    previousLatitude = latitude
    previousLongitude = longitude
    previousTimestamp = timestamp
    lastCourse = course
    return course
  }

  private fun resolve(
    latitude: Double,
    longitude: Double,
    speedMps: Double?,
    bearingDeg: Double?,
    timestamp: Long,
  ): GpsCourse? {
    // A missing speed is treated as moving: a receiver that reports no speed at all should still
    // steer the puck rather than never producing a course.
    val moving = speedMps == null || speedMps >= MIN_SPEED_MPS
    if (moving) {
      GeoMath.normalizeBearingDeg(bearingDeg)?.let { return GpsCourse(it, timestamp) }
      derivedBearingDeg(latitude, longitude, timestamp)?.let { return GpsCourse(it, timestamp) }
    }

    val retained = lastCourse ?: return null
    if (timestamp - retained.sourceTimestamp > RETENTION_MS) return null
    return retained
  }

  private fun derivedBearingDeg(latitude: Double, longitude: Double, timestamp: Long): Double? {
    val fromLatitude = previousLatitude ?: return null
    val fromLongitude = previousLongitude ?: return null
    val fromTimestamp = previousTimestamp ?: return null
    if (fromTimestamp >= timestamp) return null
    if (GeoMath.distanceMeters(fromLatitude, fromLongitude, latitude, longitude) < MIN_DISTANCE_M) {
      return null
    }
    return GeoMath.travelBearingDeg(fromLatitude, fromLongitude, latitude, longitude)
  }

  companion object {
    /** Below roughly a walking pace the reported bearing is receiver noise. */
    const val MIN_SPEED_MPS = 0.8

    /** Two fixes closer than this describe GPS jitter, not travel. */
    const val MIN_DISTANCE_M = 2.0

    /** How long a course survives without a fix that can confirm it. */
    const val RETENTION_MS = 10_000L
  }
}
